Mathlib/Order/Filter/ENNReal.lean

```diff
@@ -7,14 +7,144 @@ import Mathlib.Topology.Algebra.Order.LiminfLimsup
+
+@[simp, norm_cast] lemma isBoundedUnder_le_toReal :
+    IsBoundedUnder (· ≤ ·) f (fun i ↦ (u i : ℝ)) ↔ IsBoundedUnder (· ≤ ·) f u := by
+  simp only [IsBoundedUnder, IsBounded, eventually_map, ← coe_le_coe, NNReal.exists, coe_mk]
+  constructor
+  · rintro ⟨b, hb⟩
+    exact ⟨b.toNNReal, by simp, by filter_upwards [hb]; simp +contextual⟩
+  · rintro ⟨b, -, hb⟩
+    exact ⟨b, hb⟩
+
+@[simp, norm_cast] lemma isBoundedUnder_ge_toReal :
+    IsBoundedUnder (· ≥ ·) f (fun i ↦ (u i : ℝ)) ↔ IsBoundedUnder (· ≥ ·) f u := by
+  simp only [IsBoundedUnder, IsBounded, eventually_map, ← coe_le_coe, NNReal.exists, coe_mk]
+  constructor
+  · rintro ⟨b, hb⟩
+    exact ⟨b.toNNReal, by simp, by simpa⟩
+  · rintro ⟨b, -, hb⟩
+    exact ⟨b, hb⟩
+
+@[simp, norm_cast] lemma isCoboundedUnder_le_toReal [f.NeBot] :
+    IsCoboundedUnder (· ≤ ·) f (fun i ↦ (u i : ℝ)) ↔ IsCoboundedUnder (· ≤ ·) f u := by
+  simp only [IsCoboundedUnder, IsCobounded, eventually_map, ← coe_le_coe, NNReal.forall,
+    NNReal.exists]
+  constructor
+  · rintro ⟨b, hb⟩
+    exact ⟨b.toNNReal, by simp, fun x _ ↦ by simpa [*] using hb _⟩
+  · rintro ⟨b, hb₀, hb⟩
+    exact ⟨b, fun x hx ↦ hb _ (hx.exists.choose_spec.trans' (by simp)) hx⟩
+
+@[simp, norm_cast] lemma isCoboundedUnder_ge_toReal :
+    IsCoboundedUnder (· ≥ ·) f (fun i ↦ (u i : ℝ)) ↔ IsCoboundedUnder (· ≥ ·) f u := by
+  simp only [IsCoboundedUnder, IsCobounded, eventually_map, ← coe_le_coe, NNReal.forall,
+    NNReal.exists]
+  constructor
+  · rintro ⟨b, hb⟩
+    exact ⟨b, hb _ (by simp), fun x _ ↦ hb _⟩
+  · rintro ⟨b, hb₀, hb⟩
+    refine ⟨b, fun x hx ↦ ?_⟩
+    obtain hx₀ | hx₀ := le_total x 0
+    · exact hx₀.trans hb₀
+    · exact hb _ hx₀ hx
+
+@[simp]
+lemma limsSup_of_not_isBounded {f : Filter ℝ≥0} (hf : ¬ f.IsBounded (· ≤ ·)) : limsSup f = 0 := by
+  rw [limsSup, ← bot_eq_zero]
+  convert sInf_empty
+  simpa [Set.eq_empty_iff_forall_not_mem, IsBounded] using hf
+
+@[simp]
+lemma limsInf_of_not_isCobounded {f : Filter ℝ≥0} (hf : ¬ f.IsCobounded (· ≥ ·)) :
+    limsInf f = 0 := by rwa [limsInf, sSup_of_not_bddAbove]
+
+@[simp]
+lemma limsup_of_not_isBoundedUnder (hf : ¬ f.IsBoundedUnder (· ≤ ·) u) : limsup u f = 0 :=
+  limsSup_of_not_isBounded hf
+
+@[simp]
+lemma liminf_of_not_isCoboundedUnder (hf : ¬ f.IsCoboundedUnder (· ≥ ·) u) : liminf u f = 0 :=
+  limsInf_of_not_isCobounded hf
+
+@[simp, norm_cast]
+lemma toReal_liminf : liminf (fun i ↦ (u i : ℝ)) f = liminf u f := by
+  by_cases hf : f.IsCoboundedUnder (· ≥ ·) u; swap
+  · simp [*]
+  refine eq_of_forall_le_iff fun c ↦ ?_
+  rw [← Real.toNNReal_le_iff_le_coe, le_liminf_iff (by simpa) ⟨0, by simp⟩, le_liminf_iff]
+  simp only [← coe_lt_coe, Real.coe_toNNReal', lt_sup_iff, or_imp, isEmpty_Prop, not_lt,
+    zero_le_coe, IsEmpty.forall_iff, and_true, NNReal.forall, coe_mk, forall_swap (α := _ ≤ _)]
+  refine forall₂_congr fun r hr ↦ ?_
+  simpa using (le_or_lt 0 r).imp_right fun hr ↦ .of_forall fun i ↦ hr.trans_le (by simp)
```

Comment:
Except for Real.coe_toNNReal', this looks like the sort of thing I'd want automation to do... seeing `isEmpty_Prop` is funny

Reply:
Agreed! The precise issue here is that simp doesn't know how to case split along `le_or_lt 0 r`

Reply:
`grind` knows about case splitting... :eyes:

Reply:
Hmm